specreduce/wavelength_calibration.py

```diff
@@ -0,0 +1,248 @@
+                 fitter=LMLSQFitter(calc_uncertainties=True),
+                 default_refinement_method=None, default_refinement_kwargs={}):
+        """
+        input_spectrum: `~specutils.Spectrum1D`
+            A one-dimensional Spectrum1D calibration spectrum from an arc lamp or similar.
+        lines: str, list
+            List of lines to anchor the wavelength solution fit. List items are coerced to
+            CalibrationLine objects if passed as tuples of (wavelength, pixel) with
+            default_refinement_method and default_refinement_kwargs as defaults.
+        model: `~astropy.modeling.Model`
+            The model to fit for the wavelength solution. Defaults to a linear model.
+        default_refinement_method: str, optional
+            words
+        default_refinement_kwargs: dict, optional
+            words
+        """
+        self.input_spectrum = input_spectrum
+        self.model = model
+        self.spectral_unit = spectral_unit
+        self.fitter = fitter
+        self.default_refinement_method = default_refinement_method
+        self.default_refinement_kwargs = default_refinement_kwargs
+
+        if self.default_refinement_method in ("gaussian", "min", "max"):
+            if 'range' not in self.default_refinement_kwargs:
+                raise ValueError("You must define 'range' in default_refinement_kwargs to use "
+                                 f"{self.refinement_method} refinement.")
+        elif (self.default_refinement_method == "gradient" and 'direction' not in
+                self.default_refinement_kwargs):
+            raise ValueError("You must define 'direction' in default_refinement_kwargs to use "
+                             "gradient refinement")
+
+        self.lines = []
+        if isinstance(lines, str):
+            if lines in self.available_line_lists:
+                raise ValueError(f"Line list '{lines}' is not an available line list.")
+        else:
+            for line in lines:
+                if isinstance(line, CalibrationLine):
+                    self.lines.append(line)
+                else:
+                    self.lines.append(CalibrationLine(self.input_spectrum, line[0], line[1],
+                                      self.default_refinement_method,
+                                      self.default_refinement_kwargs))
+
+    @classmethod
+    def autoidentify(cls, input_spectrum, line_list, model=Linear1D):
+        # line_list could be a string ("common stellar") or an object
+        # this builds a list of CalibrationLine objects and passes to __init__
+        return cls(...)
+
+    @property
+    def refined_lines(self):
+        return [line.with_refined_pixel for line in self.lines]
+
+    @property
+    def refined_pixels(self):
+        # useful for plotting over spectrum to change input lines/refinement options
+        return [(line.wavelength, line.refined_pixel) for line in self.lines]
+
+    @cached_property
```

Comment:
cache needs to be cleared when appropriate

Reply:
That can be accomplished by deleting the cached properties from the class `__dict__`, right? I haven't use `cached_property` before

Reply:
exactly. In jdaviz I wrote a [helper method](https://github.com/spacetelescope/jdaviz/blob/main/jdaviz/core/template_mixin.py#L242-L250) that might also be useful in this case

Reply:
Alright, I implemented what I think you had in mind, take a look.

Reply:
that looks good to me, but I think there are other cases that need to clear too (`wcs` cache needs to clear for a change in pretty much any of the original inputs, I think).  I wonder if there's a way to generalize away this logic so we don't need to have all this boilerplate code 🤔 (definitely can generalize later - it would probably take a custom `__setattr__` which might be a little hacky-feeling)

Reply:
Also now clears the WCS cache on changing the `lines` or `model` attributes. Do you think that suffices for now? I imagine that at some point people would just make a new `WavelengthCalibration1D` rather than updating an existing one.